homeassistant/components/upcloud/__init__.py

```diff
@@ -52,65 +64,199 @@
+    async_dispatcher_send(
+        hass, _config_entry_update_signal_name(config_entry), config_entry
+    )
 
-    manager = upcloud_api.CloudManager(username, password)
+
+async def async_setup_entry(hass: HomeAssistantType, config_entry: ConfigEntry) -> bool:
+    """Set up the UpCloud config entry."""
+
+    manager = upcloud_api.CloudManager(
+        config_entry.data[CONF_USERNAME], config_entry.data[CONF_PASSWORD]
+    )
 
     try:
-        manager.authenticate()
-        hass.data[DATA_UPCLOUD] = UpCloud(manager)
+        await hass.async_add_executor_job(manager.authenticate)
     except upcloud_api.UpCloudAPIError:
-        _LOGGER.error("Authentication failed")
+        _LOGGER.error("Authentication failed", exc_info=True)
         return False
-
-    def upcloud_update(event_time):
-        """Call UpCloud to update information."""
-        _LOGGER.debug("Updating UpCloud component")
-        hass.data[DATA_UPCLOUD].update()
-        dispatcher_send(hass, SIGNAL_UPDATE_UPCLOUD)
+    except requests.exceptions.RequestException as err:
+        _LOGGER.error("Failed to connect", exc_info=True)
+        raise ConfigEntryNotReady from err
+
+    upcloud_data = hass.data.setdefault(DATA_UPCLOUD, UpCloudHassData())
+
+    # Handle pre config entry (0.117) scan interval migration to options
+    migrated_scan_interval = upcloud_data.scan_interval_migrations.pop(
+        config_entry.data[CONF_USERNAME], None
+    )
+    if migrated_scan_interval and (
+        not config_entry.options.get(CONF_SCAN_INTERVAL)
+        or config_entry.options[CONF_SCAN_INTERVAL] == DEFAULT_SCAN_INTERVAL.seconds
+    ):
+        update_interval = migrated_scan_interval
+        hass.config_entries.async_update_entry(
+            config_entry,
+            options={CONF_SCAN_INTERVAL: update_interval.seconds},
+        )
+    elif config_entry.options.get(CONF_SCAN_INTERVAL):
+        update_interval = timedelta(seconds=config_entry.options[CONF_SCAN_INTERVAL])
+    else:
+        update_interval = DEFAULT_SCAN_INTERVAL
+
+    coordinator = upcloud_data.coordinators[
+        config_entry.data[CONF_USERNAME]
+    ] = UpCloudDataUpdateCoordinator(
+        hass,
+        update_interval=update_interval,
+        cloud_manager=manager,
+        username=config_entry.data[CONF_USERNAME],
+    )
+
+    # Listen to config entry updates
+    coordinator.unsub_handlers.append(
+        config_entry.add_update_listener(_async_signal_options_update)
+    )
+    coordinator.unsub_handlers.append(
+        async_dispatcher_connect(
+            hass,
+            _config_entry_update_signal_name(config_entry),
+            coordinator.async_update_config,
+        )
+    )
 
     # Call the UpCloud API to refresh data
-    upcloud_update(dt.utcnow())
-    track_time_interval(hass, upcloud_update, scan_interval)
+    await coordinator.async_request_refresh()
+    if not coordinator.last_update_success:
+        raise ConfigEntryNotReady
```

Comment:
WHen you raise this method, you should not leave any objects around. In line 200 you store the coordinator already in hass.data (`upcloud_data`)

Reply:
Addressed.